Add: Functional tests suite (#1525)
telefonicaid/iotagent-node-lib: test/functional/README.md

> ## Functional test suite

Reviewer: The test/functional/README.md file should be linked in the root README.md file, from the right place.

Author: Added here: https://github.com/telefonicaid/iotagent-node-lib/pull/1525/commits/05b86a5e4bfecc4cce4f320b810334a4b590fc64

This directory contains the functional test suite for the IoTA JSON. This test suite is based on mocha and chai. For
mocks, we use the nock library. Additionally, it uses some specific functions to ease to implement the test. Helper
functions are located in the `testUtils.js` file.

There are 2 tests files in this directory:

-   `fuctional-tests.js`: This file contains the test defined in the "classic way". This means, coded in the JS file as
    any other mocha test. It uses the functions defined in the `testUtils.js` file to simplify the tests.

> -   `functional-tests-auto.js`: This file contains the test defined in the "automatic way". This means, the test cases

Reviewer: As discussed at Teams, a better name could be:

```suggestion
-   `functional-tests-runner.js`: This file contains the test defined in the "automatic way". This means, the test cases
```

(Adjust filename, comments, etc. where it applies along this PR)

Author: Done in https://github.com/telefonicaid/iotagent-node-lib/pull/1525/commits/e6d978198cf097d72ea7f26dcba22915e2f15554 and https://github.com/telefonicaid/iotagent-node-lib/pull/1525/commits/ac1ddf50bf17290a3a57d608e3bc6e603b645403

    are defined as JSON in a separate file (`testCases.js`). This file is loaded by the test suite and the test cases

>     are automatically generated. This is the recommended way to define the test cases.
> 

Reviewer: ```suggestion
    are automatically generated.

The recommended way is to use `testCases.js` (run by `functional-tests-runner.js`). The `fuctional-tests.js` file is provides as basically as an example in the case the "old way" needs to be used.
```

Author: Done in https://github.com/telefonicaid/iotagent-node-lib/pull/1525/commits/e6d978198cf097d72ea7f26dcba22915e2f15554 and https://github.com/telefonicaid/iotagent-node-lib/pull/1525/commits/ac1ddf50bf17290a3a57d608e3bc6e603b645403

### Automatic test cases

Each test case is defined as a JSON object in the `testCases.js` file. This file is loaded by the test suite and the
test cases are automatically generated. Each test case is defined as an object with the following elements:

-   `describeName`: The name of the `DESCRIBE` test case. This will be used to generate the test case name in the mocha
    test suite.
-   `provision`: The JSON object that will be sent to the IoTA JSON provisioning API. This will be used to create the
    group. It contains the following elements:
    -   `url`: The URL of the provisioning API (group)
    -   `method`: The HTTP method to use (POST)
    -   `json`: The JSON object that defines the group
    -   `headers`: The headers to send to the provisioning API. This should contain the `fiware-service` and
        `fiware-servicepath` headers.
    -   `skip`: optional. If set to `true`, the test case (`describe`) will be skipped. This is useful to skip test

>         cases that are not supported by the agent. It can also have a string value `lib`. This will skip the test case

Reviewer: ```suggestion
        cases that are not supported by the agent. It can also have a string value `"lib"`. This will skip the test case
```

Given that another possibility is boolean `true`, this way we make more explicit that this case referst to a string, not a boolean.

functions are located in the `testUtils.js` file.

There are 2 tests files in this directory:

-   `fuctional-tests.js`: This file contains the test defined in the "classic way". This means, coded in the JS file as
    any other mocha test. It uses the functions defined in the `testUtils.js` file to simplify the tests.
-   `functional-tests-runner.js`: This file contains the test runner that executes the tests defined as JSON in a
    separate file (`testCases.js`) in a "automatic way". This file is loaded by the test suite and the test cases are
    automatically executed.

The recommended way is to use `testCases.js` (run by `functional-tests-runner.js`). The `fuctional-tests.js` file is
provides as basically as an example in the case the "old way" needs to be used (I.E: when the test follow a different
pattern than the one supported by the test runner).

### Automatic test cases

Each test case is defined as a JSON object in the `testCases.js` file. This file is loaded by the test suite and the
test cases are automatically generated. Each test case is defined as an object with the following elements:

> -   `describeName`: The name of the `DESCRIBE` test case. This will be used to generate the test case name in the mocha

Reviewer: ```suggestion
-   `describeName`: The name of the `DESCRIBE` test case. This will be used to generate the test case name in the mocha. Note this name is prefixed by a pure number (e.g `0010 - Simple group without attributes`) which specifies the group to which the test belong (usually meaning a feature) or by a number preced by the `#` symbol to refer to an issue number (e.g. `#1234 - Bug foobar`).
```

            }
        }
    ]
}
```

### Advanced features

#### Multientity

This test suite support the multientity feature. To test this feature, you need to set add to the test case the
parameter `type: 'multientity'`. This will automatically take care of the multientity feature. This means that the suite
will configure the mock to listen to `/v2/op/update` instead of `/v2/entities?options=upsert`.

In particular, it will configure the mock to listen the correct URL. You should define the expectation for the test case
as a batch operation (see the following example).

```javascript
{
    "entities": [
        {
            "id": "TheLightType2:MQTT_2",
            "type": "TheLightType2",
            "status": {
                "value": false,
                "type": "Boolean"
            }
        },
        {
            "id": "TheLightType2:MQTT_3",
            "type": "TheLightType2",
            "temperature": {
                "value": 10,
                "type": "Number"
            }
        }
    ],
    "actionType": "append"
}
```

#### Multimeasures

It is also supported to test cases in which is sent more than one measure. To do so, you need to define the test case

> expectation as an array, with one object for each measurement. Then, the suite will recognize the array length and will

Reviewer: ```suggestion
`expectation` as an array, with one object for each measurement. Then, the suite will recognize the array length and will
```

expect the same number of NGSI requests. I.E:

```js
[
    {
        id: 'TheLightType2:MQTT_2',
        type: 'TheLightType2',
        temperature: {
            value: 10,
            type: 'Number'
        },
        status: {
            value: false,
            type: 'Boolean'
        }
    },
    {
        id: 'TheLightType2:MQTT_2',
        type: 'TheLightType2',
        temperature: {
            value: 20,
            type: 'Number'
        },
        status: {
            value: true,
            type: 'Boolean'
        }
    }
];
```

> You also should define the measure as multimeasure. This is done by defining the measure JSON element as an array of

Reviewer: ```suggestion
You also should define the measure as multimeasure. This is done by defining the `measure` JSON element as an array of
```

send the measure to the correct topic based on the `i` and `k` parameters. It will ignore the `url` and `method`
parameters present in the measure JSON element. I.E:

```javascript
should: [
    {
        transport: 'MQTT',
        shouldName: 'should send its value to the Context Broker when using MQTT',
        type: 'single',
        measure: {
            url: 'http://localhost:' + config.http.port + '/iot/json',
            method: 'POST',
            qs: {
                i: 'MQTT_2',
                k: '123456'
            },
            json: {
                s: false,
                t: 10
            }
        },
        expectation: {
            id: 'TheLightType2:MQTT_2',
            type: 'TheLightType2',
            temperature: {
                value: 10,
                type: 'Number'
            },
            status: {
                value: false,
                type: 'Boolean'
            }
        }
    }
];
```

#### No payload reception

The test suite also supports the case in which the Context Broker does not receive any payload. This is done by defining
the expectation as an empty object. I.E:

```javascript
    ...
    expectation: []
    ...
```

> 

Reviewer: ```suggestion

Note that this means the CB *must not* receive any payload. In other words, if the CB would receive any payload, the test will fail.

```

To clarify that `[]` here means "we check nothing is recieved" versus "no check is done".
